pulp_container/app/registry_api.py

```diff
@@ -326,29 +300,39 @@ def get_dr_push(self, request, path, create=False):
                 if not repository.PUSH_ENABLED:
                     raise RepositoryInvalid(name=path, message="Repository is read-only.")
             elif create:
-                try:
-                    with transaction.atomic():
-                        repo_serializer = serializers.ContainerPushRepositorySerializer(
-                            data={"name": path}, context={"request": request}
-                        )
-                        repo_serializer.is_valid(raise_exception=True)
-                        repository = repo_serializer.create(repo_serializer.validated_data)
-                        distribution.repository = repository
-                        distribution.save()
-                except IntegrityError:
-                    # Seems like another process created our stuff already. Retry fetching it.
-                    distribution = models.ContainerDistribution.objects.get(pk=distribution.pk)
-                    repository = distribution.repository
-                    if repository:
-                        repository = repository.cast()
-                        if not repository.PUSH_ENABLED:
-                            raise RepositoryInvalid(name=path, message="Repository is read-only.")
-                    else:
-                        raise RepositoryNotFound(name=path)
+                with transaction.atomic():
+                    repository = serializers.ContainerPushRepositorySerializer.get_or_create(
+                        {"name": path}
+                    )
+                    distribution.repository = repository
+                    distribution.save()
+                # Seems like another process created our stuff already. Retry fetching it.
+                distribution = models.ContainerDistribution.objects.get(pk=distribution.pk)
+                repository = distribution.repository
+                if repository:
+                    repository = repository.cast()
+                    if not repository.PUSH_ENABLED:
+                        raise RepositoryInvalid(name=path, message="Repository is read-only.")
+                else:
+                    raise RepositoryNotFound(name=path)
             else:
                 raise RepositoryNotFound(name=path)
         return distribution, repository
 
+    def create_dr(self, path, request):
+        with transaction.atomic():
+            repository = serializers.ContainerPushRepositorySerializer.get_or_create({"name": path})
+            distribution = serializers.ContainerDistributionSerializer.get_or_create(
+                {"base_path": path, "name": path}, {"repository": get_url(repository)}
+            )
+
+            if distribution.repository:
```

Comment:
it needs and else:
distribution.repository = repository
 distribution.save()

Reply:
It is not needed if we create a new distribution. We want to check the names and references without updating anything.
```
In [1]: from pulp_container.app.serializers import ContainerPushRepositorySerializer

In [2]: from pulp_container.app.serializers import ContainerDistributionSerializer

In [3]: repo = ContainerPushRepositorySerializer.get_or_create({"name": "bla"})

In [4]: from pulp_container.app.models import ContainerPushRepository, ContainerDistribution

In [5]: ContainerPushRepository.objects.filter(name="bla")
Out[5]: <QuerySet [<ContainerPushRepository (pulp_type=container.container-push): pk=018bb01f-a735-79ab-a913-8878aec9226c>]>

In [6]: ContainerDistribution.objects.filter(name="bla")
Out[6]: <QuerySet []>

In [7]: from pulpcore.plugin.util import get_url

In [8]: dist = ContainerDistributionSerializer.get_or_create({"name": "bla", "base_path": "bla"}, {"repository": get_url(repo)})

In [9]: dist
Out[9]: <ContainerDistribution (pulp_type=container.container): pk=018bb022-33d0-723e-8562-576ce14e9ed8>

In [10]: dist.repository
Out[10]: <ContainerPushRepository (pulp_type=container.container-push): pk=018bb01f-a735-79ab-a913-8878aec9226c>

In [11]: repo
Out[11]: <ContainerPushRepository (pulp_type=container.container-push): pk=018bb01f-a735-79ab-a913-8878aec9226c>
```

Reply:
The code path suggests that we are about to create a new repository:
```
            distribution = models.ContainerDistribution.objects.get(base_path=path)
        except models.ContainerDistribution.DoesNotExist:
            if create:
                distribution, repository = self.create_dr(path, request)
```
Do you agree?

Reply:
i agree that we are about to create new distribution, however to account for race condition during get_or_create ( and that's why we're using it here) another process could have created a distribution(without a repo) and the get_or_create would get the newly-created-dist by another process

Reply:
Yes, you are right!